x/gamm/keeper/migrate.go

```diff
@@ -355,3 +356,73 @@ func (k Keeper) UpdateMigrationRecords(ctx sdk.Context, records []gammmigration.
 	}
 	return nil
 }
+
+// createConcentratedPoolFromCFMM creates a new concentrated liquidity pool with the desiredDenom0 token as the
+// token 0, links it with an existing CFMM pool, and returns the created pool.
+// It uses pool manager module account as the creator of the pool.
+// Returns error if desired denom 0 is not in associated with the CFMM pool.
+// Returns error if CFMM pool does not have exactly 2 denoms.
+// Returns error if pool creation fails.
+func (k Keeper) CreateConcentratedPoolFromCFMM(ctx sdk.Context, cfmmPoolIdToLinkWith uint64, desiredDenom0 string, spreadFactor sdk.Dec, tickSpacing uint64) (poolmanagertypes.PoolI, error) {
+	cfmmPool, err := k.GetCFMMPool(ctx, cfmmPoolIdToLinkWith)
+	if err != nil {
+		return nil, err
+	}
+
+	poolmanagerModuleAcc := k.accountKeeper.GetModuleAccount(ctx, poolmanagertypes.ModuleName)
+	poolCreatorAddress := poolmanagerModuleAcc.GetAddress()
+
+	poolLiquidity := cfmmPool.GetTotalPoolLiquidity(ctx)
+	if len(poolLiquidity) != 2 {
+		return nil, types.MustHaveTwoDenomsError{NumDenoms: len(poolLiquidity)}
+	}
+
+	foundDenom0 := false
+	denom1 := ""
+	for _, coin := range poolLiquidity {
+		if coin.Denom == desiredDenom0 {
+			foundDenom0 = true
+		} else {
+			denom1 = coin.Denom
+		}
+	}
```

Comment:
nit: wouldn't this logic be cleaner with just two if statements checking each denom?

Reply:
lol yeah this feels over engineered, will change it

Reply:
Simplified here https://github.com/osmosis-labs/osmosis/pull/5895/commits/bb111c07457d3c7cc3a191acbdc5bf0be8ef5af0